in-toto-keygen: a command line utility for generating keys

isachit opened the pull request:
A CLI tool for creating key files, and dumping them with filename , filename.pub for private and public keys respectively.
General Usage:
`python in_toto_keygen.py [-p] <filename>`

**Issue Referenced** - #96 

Please note the progress:
- [x] Write the script
- [x] Add Exceptions
- [x] Add Documentation
- [x] Write the unit tests and run them
- [ ] Fix any other issues, and add other functionalities.

@lukpueh and @SantiagoTorres.

UPDATE 1: Added @lukpueh in the author field since he is the one who actually wrote the two functions I have used.
UPDATE 2: It's pretty much in place. Just removing the unnecessary history remains.

isachit committed:
Added in_toto_keygen.py

isachit committed:
Added prompt_password function

isachit committed:
Added the tests

isachit committed:
Added functionality in test_in_toto_keygen_prompt_generate_and_write_rsa_keypair

isachit committed:
Added import of password_prompt
Fixed minor error in type field of attribute string.

isachit committed:
Deleted unnecessary code

isachit commented:
I wonder, why aren't @SantiagoTorres  and @lukpueh  in the participants section. :| :/ And also, The travis CI build, along with coveralls. Hm.. Fishy.

isachit commented:
Oh yeeesss!!!! Finally!! :D

SantiagoTorres commented on test/test_in_toto_keygen.py:
```diff
@@ -0,0 +1,99 @@
+
+
+if __name__ == '__main__':
+  unittest.main(buffer=True)
```
Can you end this line with a newline? This usually confuses linecount algorithms and causes off-by-ones on diffing

isachit replied:
Okay!

SantiagoTorres commented on test/test_in_toto_keygen.py:
```diff
@@ -0,0 +1,99 @@
+"""
```
In addition to the existing tests, I'm wondering if it'd be nice to have:

1. Create a keypair, and load the public and private keys
2. Create a keypair with different length parameters and load it

isachit replied:
Alright. On it. :)

SantiagoTorres reviewed (commented):
Looking good! 👍

isachit committed:
Added functionality in in_toto_keygen.py which now also takes an integer as an input(the key length of the RSA key)
Also added the tests in test_in_toto_keygen.py regarding the new function introduced in in_toto_keygen.py

isachit committed:
Added securesystemslib import, minor bug fixes

isachit committed:
Minor bugfixes

isachit committed:
Added patch to address the password prompt

isachit committed:
Made the bits argument(which is positional) also an optional argument with default value=3072

isachit committed:
Added new line in test_in_toto_keygen.py, at the end

isachit commented:
@SantiagoTorres  I guess, this is what you suggested right? or?

SantiagoTorres commented on in_toto/in_toto_keygen.py:
```diff
@@ -14,7 +14,9 @@
-  <filename>.pub for private and public keys respectively.
+  <filename>.pub for private and public keys respectively. It also takes an
+  integer as an input, which specifies the length of the RSA key to be
+  generated. By default it is set as 3072.
```
Yeah this is good.

SantiagoTorres reviewed (commented):
Looks good to me. @lukpueh what do you think?

isachit committed:
Fixed Documentation

lukpueh commented on in_toto/in_toto_keygen.py:
```diff
@@ -0,0 +1,170 @@
+      sys.exit(0)
+
+  except Exception as e:
+    print('The following error occurred', e)
```
Please use `log.error` instead of print.

lukpueh commented on in_toto/in_toto_keygen.py:
```diff
@@ -0,0 +1,170 @@
+  in_toto_args.add_argument("bits", nargs= "?", default=3072, type=int,
+                            help="The key size, or key length, of the RSA "
+                            "key.  'bits' must be 2048, or greater, and a "
+                            "multiple of 256.", metavar="<bits>")
```
Latter is only [partially true](https://github.com/secure-systems-lab/securesystemslib/blob/master/securesystemslib/keys.py#L211-L214) also see the [default crypto library](https://github.com/secure-systems-lab/securesystemslib/blob/master/securesystemslib/settings.py#L49).
I think you can just drop the info here.

lukpueh reviewed (changes requested):
Great work!

 I wonder though why you copied over the code from `in_toto.util`. I thought we had agreed that you will try to use `securesystemslib.keys` directly, or if that's not possible, just use `in_toto.util`.

I'd rather not have the same code twice, especially since I'd like to move that particular code to `securesystemslib.keys` altogether. 

👍  for adding the key length parameter. 
I'd welcome if you changed `in_toto.util. generate_and_write_rsa_keypair` to accept an optional bits parameter and then use that function here.

isachit committed:
Used in_toto.util's functions, edited util.py to include the bits parameter, edited tests

isachit committed:
Used log.error instead of print. Edited the description for bits argument

isachit committed:
Edited test_util.py and test_in_toto_keygen.py

lukpueh commented:
Great work! Thanks for the PR, @isachit. Out of curiosity, why did you call the branch *nonuke*?

isachit commented:
@SantiagoTorres I see what you did there! :D :P Haha. There were some issues with rebasing and removing the commit history not relevant to this PR. So @SantiagoTorres  helped me with that, and while doing it, he made the branch nonuke.

SantiagoTorres commented:
[¯\\_(ツ)_/¯](http://i.imgur.com/7A2QTxW.gif)
